### PR TITLE
8339307: jhsdb jstack could not trace FFM upcall frame

### PR DESCRIPTION
**Pull request by YaSuenag**
This PR is successor of #20789 . I got some comments in there, then I needed to fix in many pooints. And also I have to fix branch name to kick GHA automatically (#20789 is named with `pr/`, it meets the condition to skip GHA). Hence I've opened another PR for this JBS issue.

This PR has been updated with about topics since #20789:
* Use `JavaFrameAnchor` instead of raw frame pointer to unwind frame of `UpcallStub`.
* The change happens x86 (includes AMD64), aarch64, PPC64, RISC-V 64 only - s390 is out of scope because SA does not have s390 implementation.
    * Only both AMD64 and aarch64 have tested on GHA.
* Refactor testcase to meet expected condition certainly.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8339307: jhsdb jstack could not trace FFM upcall frame`

### Issue
 * [JDK-8339307](https://bugs.openjdk.org/browse/JDK-8339307): jhsdb jstack could not trace FFM upcall frame (**Bug** - P4)

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**) Review applies to [f22c6236](https://git.openjdk.org/jdk/pull/20885/files/f22c6236098b4a4fa99d5fb8668566b4c34ac5a4)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20885/head:pull/20885` \
`$ git checkout pull/20885`

Update a local copy of the PR: \
`$ git checkout pull/20885` \
`$ git pull https://git.openjdk.org/jdk.git pull/20885/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20885`

View PR using the GUI difftool: \
`$ git pr show -t 20885`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20885.diff">https://git.openjdk.org/jdk/pull/20885.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20885#issuecomment-2334084499)

### CODE REVIEW

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/ppc64/PPC64Frame.java [hunk 1]
```diff
@@ -314,6 +314,33 @@ private Frame senderForEntryFrame(PPC64RegisterMap map) {
...
+  private Frame senderForUpcallStub(PPC64RegisterMap map, UpcallStub stub) {
+    if (DEBUG) {
+      System.out.println("senderForUpcallStub");
+    }
+    if (Assert.ASSERTS_ENABLED) {
+      Assert.that(map != null, "map must be set");
+    }
+
+    var lastJavaFP = stub.getLastJavaFP(this); // This will be null
+    var lastJavaSP = stub.getLastJavaSP(this);                              var lastJavaPC = stub.getLastJavaPC(this);
```
Comment (JornVernee) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/ppc64/PPC64Frame.java:326]:
```suggestion
    var lastJavaSP = stub.getLastJavaSP(this);
    var lastJavaPC = stub.getLastJavaPC(this);
```

File: test/hotspot/jtreg/serviceability/sa/libupcall.c [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) 2014, Oracle and/or its affiliates. All rights reserved.
```
Comment (JornVernee) [inline test/hotspot/jtreg/serviceability/sa/libupcall.c:2]:
```suggestion
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
```

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/UpcallStub.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+import sun.jvm.hotspot.utilities.Observer;
+
+public class UpcallStub extends RuntimeBlob {
+
+  private static CIntegerField frameDataOffsetField;
+
+  private static AddressField lastJavaFPField;
+
+  private static AddressField lastJavaSPField;
+
```
Comment (plummercj) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/UpcallStub.java:41]: I think it's better without the blank lines.

**Review by plummercj — APPROVED** [review]
Overall the SA code looks good, but since I don't understand how UpcallStub linkage works, it's hard for me to say for sure if it is correct for what you are trying to accomplish. However, the test coverage seems to sufficiently ensure correctness of the implementation. Thanks for adding the code to have the LingeredApp readiness wait until the upcall thread is ready.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/UpcallStub.java (hunk 3 above)
Comment (YaSuenag) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/UpcallStub.java]: Removed them in new commit.

**Comment by YaSuenag** [comment]
@JornVernee @plummercj 

Thanks a lot for your review! I think the PR is ready to merge now. However `ready` label has been removed due to new commit (removing blank lines commented by Chris). So can you approve this change again? Thanks!

**Review by plummercj — APPROVED** [review]

**Pull request closed without merging**